File: backend/app/api/pagination.py
"""Canonical pagination for list endpoints.

Every paginated list endpoint shares ONE contract so the API (and the
frontend's shared Load-More control) behaves identically everywhere:

- query params ``page`` (1-based) and ``page_size``
- ``page_size`` defaults to :data:`DEFAULT_PAGE_SIZE` and is capped at
  :data:`MAX_PAGE_SIZE`
- the response envelope always carries ``items``, ``total``, ``page`` and
  ``page_size``

Use the :func:`pagination_params` dependency for the inputs and either
:class:`PageMeta` (mixed into a typed Pydantic list response) or
:func:`paginated` (for dict-returning handlers) for the envelope. Bounds
live here and nowhere else, so a new endpoint cannot quietly drift onto a
different default or cap.
"""

from dataclasses import dataclass
from typing import Any

from fastapi import Query
from pydantic import BaseModel

# The single source of truth for list-page sizing. The frontend's shared
# Load-More control requests the same DEFAULT_PAGE_SIZE, so a bare
# `GET /api/<list>` and the UI's first page return the same rows.
DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100


@dataclass(frozen=True)
class PaginationParams:
    """Resolved, validated pagination inputs for a list request."""

    page: int
    page_size: int

    @property
    def offset(self) -> int:
        return (self.page - 1) * self.page_size

    @property
    def limit(self) -> int:
        return self.page_size


def pagination_params(
    page: int = Query(1, ge=1),
    page_size: int = Query(DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE),
) -> PaginationParams:
    """FastAPI dependency yielding the canonical, bounds-checked params."""
    return PaginationParams(page=page, page_size=page_size)


class PageMeta(BaseModel):
    """Pagination fields every typed list response includes.

    Mix into a Pydantic list response alongside its ``items`` / ``total``
    so the envelope matches the dict-returning handlers byte-for-byte::

        class VendorListResponse(PageMeta):
            items: list[VendorResponse]
            total: int
    """

    page: int = 1
    page_size: int = DEFAULT_PAGE_SIZE


def paginated(items: list[Any], total: int, p: PaginationParams) -> dict[str, Any]:
    """Build the canonical envelope for a dict-returning list handler."""
    return {"items": items, "total": total, "page": p.page, "page_size": p.page_size}


# Cap on how many ids a "select all N matching" bulk-selection query ever
# returns in one call. The bulk mutation endpoints (invoice bulk delete /
# status-change, exception bulk resolve, expense bulk-gl-code) only accept an
# explicit id list — a client-side "select all" that only captured the
# currently-LOADED page silently skipped every row past it, with no warning.
# The fix is a client that resolves the *whole filtered set* of ids before
# selecting, via the `/ids` sibling of each list endpoint below — but an
# unbounded query on a huge tenant would turn "select all" into "load the
# whole table" and a single POST body with tens of thousands of ids. Capped
# here; `MatchingIdsResponse.truncated` tells the caller when it was hit, so a
# partial selection is never silently presented as complete.
MAX_SELECT_ALL_IDS = 5000


class MatchingIdsResponse(BaseModel):
    """Every id matching the caller's list filters (up to
    :data:`MAX_SELECT_ALL_IDS`), for a "select all N matching" bulk-selection
    affordance. Mirrors ``PageMeta``'s role: one shared shape so a new
    ``/ids`` sibling endpoint can't quietly drift from the others."""

    ids: list[str]
    total: int
    truncated: bool

File: frontend/src/lib/utils/pagination.ts
/**
 * Load-more append for offset-paginated lists.
 *
 * Append the next page onto the already-loaded rows, skipping any id already
 * present. Offset pagination can re-surface a row when the underlying set
 * shifts between fetches (e.g. a new invoice inserted, a notification arriving
 * between page loads); a duplicate id would crash the keyed
 * `{#each ... (item.id)}` rendering the list (Svelte 5 `each_key_duplicate`).
 *
 * Semantics (matching the original invoice-store inline guard):
 * - The existing row wins — an incoming duplicate is dropped, never replacing
 *   the already-rendered row.
 * - Order is preserved: existing rows first (in order), then the incoming
 *   page's new rows in their server order.
 */
export function appendUnique<T extends { id: string }>(existing: T[], incoming: T[]): T[] {
	const seen = new Set(existing.map((item) => item.id));
	return [...existing, ...incoming.filter((item) => !seen.has(item.id))];
}

/**
 * Response shape of every list endpoint's `/ids` sibling (`GET
 * /api/invoices/ids`, `/api/exceptions/ids`, `/api/expenses/ids` —
 * `backend/app/api/pagination.py::MatchingIdsResponse`).
 *
 * Backs the "select all N matching" affordance: a list page's header
 * checkbox only ever selects the currently-LOADED page of rows (the bulk
 * endpoints take an explicit id list, and the page only has what it fetched
 * client-side) — this is the whole filtered set instead, resolved
 * server-side so a bulk action can't silently skip rows past the first page.
 * `truncated` is true when the match count exceeded the server's cap
 * (`MAX_SELECT_ALL_IDS`), so a partial selection is never presented as
 * complete.
 */
export interface MatchingIdsResponse {
	ids: string[];
	total: number;
	truncated: boolean;
}

/** Envelope every paginated list endpoint returns (`backend/app/api/pagination.py`). */
export interface PagedResponse<T> {
	items: T[];
	total: number;
}

/**
 * The server's `MAX_PAGE_SIZE` (`backend/app/api/pagination.py`). A request for
 * more is clamped, so asking for more is not a way to skip paging.
 */
export const MAX_PAGE_SIZE = 100;

/**
 * Fetch EVERY page of an offset-paginated list, not just the first.
 *
 * A bare `api.get('/api/vendors')` returns the server's DEFAULT_PAGE_SIZE (20)
 * rows and nothing says so. That is right for a table with a Load-more control
 * and wrong for a `<select>` whose options ARE the set of valid choices: on a
 * tenant with more suppliers than one page, the dropdown silently omits most of
 * them and — with no search inside a native `<select>` — they are unreachable.
 *
 * Raising `page_size` is not the fix; the server caps it at
 * {@link MAX_PAGE_SIZE}, so it only moves the cliff. This walks pages until the
 * envelope's own `total` is satisfied.
 *
 * `maxPages` is a runaway bound, not a product limit: it stops a mis-specified
 * endpoint (a `total` that never converges) from looping forever. Reaching it
 * means the list is far past what any `<select>` should hold, and the caller
 * should have a searchable picker instead.
 */
export async function fetchAllPages<T>(
	fetchPage: (page: number, pageSize: number) => Promise<PagedResponse<T>>,
	{ pageSize = MAX_PAGE_SIZE, maxPages = 50 }: { pageSize?: number; maxPages?: number } = {}
): Promise<T[]> {
	let items: T[] = [];
	for (let page = 1; page <= maxPages; page++) {
		const data = await fetchPage(page, pageSize);
		items = items.concat(data.items ?? []);
		// Stop on the envelope's own count, and on a short/empty page so a
		// `total` that disagrees with reality can't spin.
		if (items.length >= (data.total ?? 0) || (data.items?.length ?? 0) < pageSize) break;
	}
	return items;
}
